src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs

```diff
@@ -173,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
 
             try
             {
-                // This variable records any error occurring in the thread pool task's context.
-                Exception error = null;
-
-                await Task.Run(() =>
-                {
-                    try
-                    {
-                        this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).Wait(cancellation);
-                    }
-                    catch (Exception e)
-                    {
-                        // Record the error occurring in the thread pool's context.
-                        error = e;
-                    }
-                }).ConfigureAwait(false);
-
-                // Throw the error within this error handling context.
-                if (error != null)
-                    throw error;
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttempt(false, endPoint));
+                await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);
```

Comment:
`WithCancellationAsync` replace `.Wait(cancellation)`

Reply:
pardon?
the core of this change is to not block (always discouraged) and WithCancellationAsync extension solve exactly this problem (now you can await with a timeout)

In the extension comment I left a link to a microsoft blog post that explain this.